src/v/storage/segment_index.h

```diff
@@ -71,6 +83,8 @@ class segment_index {
     ss::future<> flush();
     ss::future<> truncate(model::offset);
 
+    ss::future<ss::file> open();
```

Comment:
I do not understand the semantics of this new `open()` method.

In a few places we just call open without closing the resulting file, but as far as I can tell one invariant is that every  `ss::file` object that is opened/created should be closed.

It seems that `ss::file` is internally reference counted, is that what we are relying on? I'm not sure about this since it seems like this is only for read-only files shared across fibers, and in any case I don't think it removes the need to call `close()`. Most of my other comments in this and the .cc file are related to this.

Reply:
Just to clear, the semantics of `ss:file` appear to roughly match the semantics of a raw `int` fd. That is, copying a file object `f` like `auto f2 = f` makes both objects point to the same underlying fd: only one should be closed. OTOH using `f.dup()` like `auto f2 = f.dup()` makes each object _logically_ point to a different rd, though implementation-wise they actually point at the _same_ fd but using reference counting internally to call `close(2)` only when the last object pointing to the fd is actually destroyed.

Reply:
In the case of segment_index, the rules for this file object are meant to be simple: we only grab one while doing a materialize/flush/truncate, and we close it when we're done.  The trouble was I hadn't written the code that way.  I've fixed this with proper open->use->close code in these places.

Reply:
Thanks John, it makes sense.